### PR TITLE
Add more validations to Metrics tests

### PR DESCRIPTION
**Pull request by cijothomas**
Still tons of duplication. Not yet in a position to fully extract common code. That'll be done once I am done with adding all scenarios.

This PR adds test coverage to validate delta/cumulative is done correctly by reporting more measurements after the 1st collect.
Also added a basic test for Gauge

### CODE REVIEW

File: opentelemetry-sdk/src/metrics/mod.rs [hunk 1]
```diff
@@ -1168,6 +1312,38 @@ mod tests {
         let data_point1 = find_datapoint_with_key_value(&sum.data_points, "key1", "value2")
             .expect("datapoint with key1=value2 expected");
         assert_eq!(data_point1.value, 3);
+
+        // Reset and report more measurements
+        test_context.reset_metrics();
+        counter.add(1, &[KeyValue::new("key1", "value1")]);
```
Comment (utpilla) [inline opentelemetry-sdk/src/metrics/mod.rs:1318]: UpDownCounter should be tested with negative measurements as well.

File: opentelemetry-sdk/src/metrics/mod.rs [hunk 2]
```diff
@@ -1046,39 +1053,144 @@ mod tests {
...
+            assert_eq!(data_point1.count, values_kv2.len() as u64);
+            assert_eq!(data_point1.sum, values_kv2.iter().sum::<u64>());
+            assert_eq!(data_point1.min.unwrap(), *values_kv2.iter().min().unwrap());
+            assert_eq!(data_point1.max.unwrap(), *values_kv2.iter().max().unwrap());
+        }
+    }
+
+    fn gauge_aggregation_helper() {
+        // Arrange
+        let mut test_context = TestContext::new(Temporality::Delta);
```
Comment (utpilla) [inline opentelemetry-sdk/src/metrics/mod.rs:1139]: We should test this for both `Delta` and `Cumulative` to ensure that Gauge always uses last value aggregation regardless of the aggregation temporality of the `MetricReader`.

**Review by utpilla — APPROVED** [review]

File: opentelemetry-sdk/src/metrics/mod.rs [hunk 3]
```diff
@@ -1046,39 +1053,144 @@ mod tests {
...
+        if temporality == Temporality::Cumulative {
+            assert_eq!(data_point1.count, 2 * (values_kv2.len() as u64));
+            assert_eq!(data_point1.sum, 2 * (values_kv2.iter().sum::<u64>()));
+            assert_eq!(data_point1.min.unwrap(), *values_kv2.iter().min().unwrap());
+            assert_eq!(data_point1.max.unwrap(), *values_kv2.iter().max().unwrap());
+        } else {
+            assert_eq!(data_point1.count, values_kv2.len() as u64);
+            assert_eq!(data_point1.sum, values_kv2.iter().sum::<u64>());
+            assert_eq!(data_point1.min.unwrap(), *values_kv2.iter().min().unwrap());
+            assert_eq!(data_point1.max.unwrap(), *values_kv2.iter().max().unwrap());
```
Comment (lalitb) [inline opentelemetry-sdk/src/metrics/mod.rs:1133]: Thanks, tests look more complete now :)

File: opentelemetry-sdk/src/metrics/mod.rs (hunk 2 above)
Comment (lalitb) [reply opentelemetry-sdk/src/metrics/mod.rs:1139]: +1

**Review by lalitb — APPROVED** [review]

**Pull request merged**